src/docker/blockchain/ethereum.rs

```diff
@@ -60,47 +63,107 @@ impl Image for EthereumNode {
     }
+
+    Ok(contract_address)
+}
+
+async fn deploy_erc20_contract(client: Web3<Http>) -> Result<Address, web3::Error> {
+    let data = TOKEN_CONTRACT[2..].trim(); // remove the 0x in the front and any whitespace
+    let erc20_contract = hex::decode(data).expect("token contract should be valid hex");
+
+    let receipt = send_transaction(client, None, 10_000_000, U256::from(0), erc20_contract).await?;
+
+    let contract_address = receipt
+        .contract_address
+        .expect("we deployed a contract, should have contract address");
+
+    Ok(contract_address)
+}
+
+fn transfer_fn(address: Address, amount: U256) -> Result<Vec<u8>, ethabi::Error> {
+    ethabi::Contract::load(Cursor::new(CONTRACT_ABI))?
+        .function("transfer")?
+        .encode_input(&[Token::Address(address), Token::Uint(amount)])
+}
+
+async fn send_transaction(
+    client: Web3<Http>,
+    to: Option<Address>,
+    gas_limit: u64,
+    amount: U256,
+    data: Vec<u8>,
+) -> Result<TransactionReceipt, web3::Error> {
+    let chain_id = get_chain_id(client.clone()).await?;
+    let tx = emerald_rs::Transaction {
+        nonce: client
+            .eth()
+            .transaction_count(*DEV_ACCOUNT, None)
+            .compat()
+            .await?
+            .as_u64(),
+        gas_price: [0u8; 32],
+        gas_limit,
+        to: to.map(|a| emerald_rs::Address(a.0)),
+        value: amount.into(),
+        data,
+    };
+    let signed_raw_tx = tx
+        .to_signed_raw(*DEV_ACCOUNT_PRIVATE_KEY, chain_id)
+        .expect("signing transaction should work");
+
+    client
+        .send_raw_transaction_with_confirmation(Bytes(signed_raw_tx), Duration::from_millis(100), 1)
+        .compat()
+        .await
+}
+
+async fn get_chain_id(client: Web3<Http>) -> Result<u8, web3::Error> {
+    let network = client.net().version().compat().await?;
+    network.parse::<u8>().map_err(|e| {
+        web3::Error::InvalidResponse(format!(
+            "{} is not a valid chain id because it cannot be parsed as a u8: {:?}",
+            network, e
+        ))
+    })
 }
 
 pub fn derive_address(secret_key: secp256k1::SecretKey) -> Address {
-    let public_key =
-        secp256k1::PublicKey::from_secret_key(&secp256k1::Secp256k1::new(), &secret_key);
-
-    let serialized_public_key = public_key.serialize_uncompressed();
-    // Remove the silly openssl 0x04 byte from the front of the
-    // serialized public key. This is a bitcoin thing that
-    // ethereum doesn't want. Eth pubkey should be 32 + 32 = 64 bytes.
-    let actual_public_key = &serialized_public_key[1..];
-    let hash = tiny_keccak::keccak256(actual_public_key);
-    // Ethereum address is the last twenty bytes of the keccak256 hash
-    let ethereum_address_bytes = &hash[12..];
-    let mut address = Address::default();
-    address.assign_from_slice(ethereum_address_bytes);
-    address
+    let address = PrivateKey::from(secret_key)
+        .to_address()
+        .expect("can never happen");
```

Comment:
Because it was using an old version of secp256k1 where conversions still returned errors and I didn't want to touch the API definition during my upgrade.